src/platforms/linux/daemon/dbusservice.cpp

```diff
@@ -130,11 +136,162 @@ QString DBusService::getLogs() {
+    if (!rv) {
+      break;
+    }
+
+    /* The set of return codes that are "permanent failures". All other
+     * possibilities are potentially transient.
+     *
+     * This is according to https://sourceware.org/glibc/wiki/NameResolver which
+     * states: "From the perspective of the application that calls getaddrinfo()
+     * it perhaps doesn't matter that much since EAI_FAIL, EAI_NONAME and
+     * EAI_NODATA are all permanent failure codes and the causes are all
+     * permanent failures in the sense that there is no point in retrying
+     * later."
+     *
+     * So this is what we do, except FreeBSD removed EAI_NODATA some time ago,
+     * so that's conditional.
+     */
+    if (rv == EAI_NONAME || rv == EAI_FAIL ||
+#ifdef EAI_NODATA
+        rv == EAI_NODATA ||
+#endif
+        (retries >= 0 && !retries--)) {
+      logger.log() << "Failed to resolve the address endpoint";
+      return false;
+    }
+
+    logger.log() << "Trying again in" << (timeout / 1000000.0) << "seconds";
+    usleep(timeout);
+  }
+
+  if ((resolved->ai_family == AF_INET &&
+       resolved->ai_addrlen == sizeof(struct sockaddr_in)) ||
+      (resolved->ai_family == AF_INET6 &&
+       resolved->ai_addrlen == sizeof(struct sockaddr_in6))) {
+    memcpy(endpoint, resolved->ai_addr, resolved->ai_addrlen);
+    freeaddrinfo(resolved);
+    return true;
+  }
+
+  logger.log() << "Invalid endpoint" << address;
+  freeaddrinfo(resolved);
+  return false;
+}
+
+bool DBusService::switchServer(const Config& config) {
+  logger.log() << "Switching server";
+
+  wg_device* device = static_cast<wg_device*>(calloc(1, sizeof(*device)));
+  if (!device) {
+    logger.log() << "Allocation failure";
+    return false;
+  }
+
+  auto guard = qScopeGuard([&] { wg_free_device(device); });
+
+  strncpy(device->name, WG_INTERFACE, IFNAMSIZ - 1);
+  device->name[IFNAMSIZ - 1] = '\0';
+
+  wg_peer* peer = static_cast<wg_peer*>(calloc(1, sizeof(*peer)));
+  if (!peer) {
+    logger.log() << "Allocation failure";
+    return false;
+  }
+
+  peer->flags =
+      (wg_peer_flags)(WGPEER_HAS_PUBLIC_KEY | WGPEER_REPLACE_ALLOWEDIPS);
+
+  device->first_peer = device->last_peer = peer;
+  device->flags = WGDEVICE_REPLACE_PEERS;
+
+  if (!parseEndpoint(&peer->endpoint.addr, config.m_serverIpv4AddrIn,
+                     config.m_serverPort)) {
+    return false;
+  }
+
+  wg_key_from_base64(peer->public_key, config.m_serverPublicKey.toLocal8Bit());
+
+  for (const IPAddressRange& ip : config.m_allowedIPAddressRanges) {
+    wg_allowedip* allowedip =
+        static_cast<wg_allowedip*>(calloc(1, sizeof(*allowedip)));
+    if (!allowedip) {
+      logger.log() << "Allocation failure";
+      return false;
+    }
+
+    if (!peer->first_allowedip) {
+      peer->first_allowedip = allowedip;
+    } else {
+      peer->last_allowedip->next_allowedip = allowedip;
+    }
+
+    peer->last_allowedip = allowedip;
+    allowedip->cidr = ip.range();
```

Comment:
c++ question - any reason to put this call after assignment to the peer->allowed_ips instead of before?

Reply:
None. We can change the order.

Reply:
No need. Just learning.